components/feature/pwa/src/main/java/mozilla/components/feature/pwa/ManifestStorage.kt

```diff
@@ -0,0 +1,42 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.pwa
+
+import android.content.Context
+import android.content.Context.MODE_PRIVATE
+import androidx.core.content.edit
+import mozilla.components.concept.engine.manifest.WebAppManifest
+import mozilla.components.concept.engine.manifest.WebAppManifestParser
+
+class ManifestStorage(context: Context) {
+    private val preferences = context.getSharedPreferences(PREFERENCE_NAME, MODE_PRIVATE)
+    private val parser = WebAppManifestParser()
+
+    fun loadManifest(startUrl: String): WebAppManifest? {
+        return preferences.getString(startUrl, null)?.let { json ->
+            when (val result = parser.parse(json)) {
+                is WebAppManifestParser.Result.Success -> result.manifest
+                is WebAppManifestParser.Result.Failure -> null
+            }
+        }
+    }
+
+    fun saveManifest(manifest: WebAppManifest) {
+        preferences.edit {
+            val json = parser.serialize(manifest).toString()
+            putString(manifest.startUrl, json)
```

Comment:
I wonder if we should *not* use the startUrl here and instead generate a UUID for this installation. This ID we will also pass to the launcher intent.
The startUrl may change over time and we probably do not want to hard-code it in the intent. And it may not really be a unique identifier. 🤔

Reply:
If we don't hard-code the startUrl in the intent, then there is no way to launch a normal browser session if we lose the manifest data. Do other browsers let a PWA migrate to a different URL?